File: src/ts/test/otel-adversarial.test.mjs
import test from "node:test";
import assert from "node:assert/strict";

import { defaultConfig } from "../dist/index.js";
import {
  createLogger,
  createOresOtelMiddleware,
  getLogContext,
  requestLogger
} from "../dist/otel.js";

const delay = (milliseconds) => new Promise((resolve) => setTimeout(resolve, milliseconds));
const settleDetachedLogs = () => delay(25);

function testConfig(timeoutMs = 5_000) {
  const config = defaultConfig("middleware-adversarial-test");
  config.environment = "test";
  config.settings.timeoutMs = timeoutMs;
  config.settings.tls.mode = "disabled";
  config.settings.tls.requireHttps = false;
  config.settings.rateLimit.enabled = false;
  config.settings.idempotency.enabled = false;
  config.settings.compression.enabled = false;
  return config;
}

function memoryLoggers(records) {
  const transport = {
    name: "memory",
    write(record) {
      records.push(record);
    }
  };
  return {
    root: createLogger({
      appName: "middleware-adversarial-test",
      name: "server",
      console: false,
      transports: transport
    }),
    fileLogger: createLogger({
      appName: "middleware-adversarial-test",
      name: "handler",
      console: false,
      transports: transport
    })
  };
}

function requestFor(slot, extraHeaders = {}) {
  return new Request(`http://example.test/orders/${slot}`, {
    headers: {
      "x-request-id": `request-${slot}`,
      "x-test-slot": String(slot),
      "x-test-user": `user-${slot}`,
      "x-test-tenant": `tenant-${slot}`,
      traceparent: `00-${Number(slot).toString(16).padStart(32, "0")}-0123456789abcdef-01`,
      ...extraHeaders
    }
  });
}

test("parallel requests never cross-contaminate request, user, tenant, or baggage context", async () => {
  const records = [];
  const { root, fileLogger } = memoryLoggers(records);
  const middleware = createOresOtelMiddleware(testConfig(), {
    logger: root,
    authVerifier: async (request) => ({
      userId: request.headers.get("x-test-user"),
      tenantId: request.headers.get("x-test-tenant"),
      claims: {
        "otel.slot": request.headers.get("x-test-slot"),
        authorization: "must-not-propagate"
      }
    })
  });

  const requestCount = 48;
  const responses = await Promise.all(
    Array.from({ length: requestCount }, async (_, slot) =>
      middleware(requestFor(slot), async (request) => {
        await delay((requestCount - slot) % 7);
        const context = getLogContext();
        assert.equal(context?.fields?.["request.id"], `request-${slot}`);
        assert.equal(context?.fields?.["user.id"], `user-${slot}`);
        assert.equal(context?.fields?.["tenant.id"], `tenant-${slot}`);
        assert.equal(context?.fields?.["baggage.otel.slot"], String(slot));

        await fileLogger.info(`file:${slot}`).send();
        const scopedLogger = requestLogger(request);
        assert.ok(scopedLogger);
        await scopedLogger.warn(`request:${slot}`).send();
        return new Response("ok", { status: 200 + (slot % 3) });
      })
    )
  );

  await settleDetachedLogs();
  assert.equal(responses.length, requestCount);
  assert.equal(getLogContext(), undefined);

  for (let slot = 0; slot < requestCount; slot += 1) {
    for (const message of [`file:${slot}`, `request:${slot}`]) {
      const matches = records.filter((record) => record.message === message);
      assert.equal(matches.length, 1, `expected one ${message} record`);
      const [record] = matches;
      assert.equal(record.fields["request.id"], `request-${slot}`);
      assert.equal(record.fields["user.id"], `user-${slot}`);
      assert.equal(record.fields["tenant.id"], `tenant-${slot}`);
      assert.equal(record.fields["baggage.otel.slot"], String(slot));
      assert.equal(record.loggedInUser?.id, `user-${slot}`);
      assert.doesNotMatch(JSON.stringify(record), /must-not-propagate|authorization/i);
    }
  }
});

test("sealed Request objects use the WeakMap logger fallback", async () => {
  const records = [];
  const { root } = memoryLoggers(records);
  const middleware = createOresOtelMiddleware(testConfig(), {
    logger: root,
    authVerifier: async () => ({ userId: "sealed-user", tenantId: "sealed-tenant" })
  });
  const request = requestFor(11, {
    "x-request-id": "request-sealed",
    traceparent: "00-aaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaa-0123456789abcdef-01"
  });
  Object.preventExtensions(request);

  const response = await middleware(request, async (scopedRequest) => {
    assert.equal(Object.hasOwn(scopedRequest, "log"), false);
    const scopedLogger = requestLogger(scopedRequest);
    assert.ok(scopedLogger);
    await scopedLogger.info("sealed request reached").send();
    return new Response(null, { status: 204 });
  });

  await settleDetachedLogs();
  assert.equal(response.status, 204);
  const record = records.find((candidate) => candidate.message === "sealed request reached");
  assert.equal(record?.fields["request.id"], "request-sealed");
  assert.equal(record?.loggedInUser?.id, "sealed-user");
  assert.equal(getLogContext(), undefined);
});

test("malformed correlation identifiers are replaced and unsafe claims are dropped", async () => {
  const records = [];
  const { root } = memoryLoggers(records);
  const middleware = createOresOtelMiddleware(testConfig(), {
    logger: root,
    authVerifier: async () => ({
      userId: "user-safe",
      tenantId: "tenant-safe",
      claims: {
        "otel.vendor": "allowed",
        authorization: "Bearer must-not-propagate",
        cookie: "must-not-propagate"
      }
    })
  });

  let observed;
  const response = await middleware(
    new Request("http://example.test/malformed", {
      headers: {
        "x-request-id": "bad request id with spaces",
        traceparent: "00-not-a-valid-trace-id-not-a-span-id-01"
      }
    }),
    async (request) => {
      observed = getLogContext();
      const scopedLogger = requestLogger(request);
      assert.ok(scopedLogger);
      await scopedLogger.info("malformed context replaced").send();
      return new Response("ok");
    }
  );

  await settleDetachedLogs();
  assert.equal(response.status, 200);
  const requestId = observed?.fields?.["request.id"];
  const traceId = observed?.fields?.["trace.id"];
  assert.equal(typeof requestId, "string");
  assert.notEqual(requestId, "bad request id with spaces");
  assert.match(requestId, /^[A-Za-z0-9._-]{1,128}$/);
  assert.match(traceId, /^[0-9a-f]{32}$/);
  assert.equal(response.headers.get("x-request-id"), requestId);
  assert.equal(response.headers.get("traceparent"), `00-${traceId}-0000000000000000-01`);

  const record = records.find((candidate) => candidate.message === "malformed context replaced");
  assert.equal(record?.fields["baggage.otel.vendor"], "allowed");
  assert.doesNotMatch(JSON.stringify(record), /Bearer must-not-propagate|cookie|authorization/i);
  assert.equal(getLogContext(), undefined);
});

test("handler failure preserves the 500 response and clears ambient context", async () => {
  const records = [];
  const { root } = memoryLoggers(records);
  const middleware = createOresOtelMiddleware(testConfig(), {
    logger: root,
    authVerifier: async () => ({ userId: "failure-user", tenantId: "failure-tenant" })
  });

  const response = await middleware(requestFor(12), async () => {
    throw new Error("handler exploded");
  });

  await settleDetachedLogs();
  assert.equal(response.status, 500);
  assert.ok(records.some((record) => record.message.startsWith("request handler failed")));
  assert.equal(records.some((record) => record.message === "request handler completed"), false);
  assert.equal(getLogContext(), undefined);
});

test("throwing transport diagnostics cannot prevent the handler or alter its response", async () => {
  let handlerRan = false;
  let transportErrors = 0;
  const root = createLogger({
    appName: "middleware-adversarial-test",
    console: false,
    transports: {
      name: "always-fails",
      write() {
        throw new Error("sink unavailable");
      }
    },
    onTransportError() {
      transportErrors += 1;
      throw new Error("diagnostic callback failed");
    }
  });
  const middleware = createOresOtelMiddleware(testConfig(), {
    logger: root,
    authVerifier: async () => ({ userId: "transport-user" })
  });

  const response = await middleware(requestFor(13), async () => {
    handlerRan = true;
    return new Response(null, { status: 204 });
  });

  await settleDetachedLogs();
  assert.equal(handlerRan, true);
  assert.equal(response.status, 204);
  assert.ok(transportErrors >= 2);
  assert.equal(getLogContext(), undefined);
});

test("a handler that outlives its deadline emits timeout rather than late completion", async () => {
  const records = [];
  const { root } = memoryLoggers(records);
  const middleware = createOresOtelMiddleware(testConfig(15), {
    logger: root,
    authVerifier: async () => ({ userId: "timeout-user", tenantId: "timeout-tenant" })
  });

  const response = await middleware(requestFor(14), async () => {
    await delay(60);
    return new Response(null, { status: 204 });
  });

  assert.equal(response.status, 504);
  await delay(80);
  await settleDetachedLogs();
  assert.ok(records.some((record) => record.message === "request handler timed out"));
  assert.equal(records.some((record) => record.message === "request handler completed"), false);
  assert.equal(getLogContext(), undefined);
});

File: src/elixir/test/otel_adversarial_test.exs
defmodule OresMiddleware.OTelAdversarialTest do
  use ExUnit.Case, async: false

  import Plug.Conn, only: [assign: 3, send_resp: 3]
  import Plug.Test, only: [conn: 2]

  alias ORESoftware.NextLoggers
  alias OresMiddleware.{Context, OTel, OTelPlug}

  setup do
    previous_context = Context.current()
    previous_metadata = Logger.metadata()

    Context.clear()
    Logger.reset_metadata([])
    log_restore = NextLoggers.clear_context()

    on_exit(fn ->
      case previous_context do
        nil -> Context.clear()
        value -> Context.put(value)
      end

      Logger.reset_metadata(previous_metadata)
      NextLoggers.restore_context(log_restore)
    end)

    :ok
  end

  test "64 concurrent request processes keep request, user, tenant, and baggage isolated" do
    logger = OTel.new_logger("middleware-adversarial-test", name: "orders")

    results =
      0..63
      |> Task.async_stream(
        fn slot ->
          slot = Integer.to_string(slot)
          context = context(slot)
          request_logger = OTel.request_logger(logger, context)

          result =
            Context.run(context, fn ->
              assert Context.current().request_id == "request-#{slot}"
              assert NextLoggers.current_context().fields["request.id"] == "request-#{slot}"
              assert NextLoggers.current_context().fields["user.id"] == "user-#{slot}"
              assert NextLoggers.current_context().fields["tenant.id"] == "tenant-#{slot}"

              file_record = NextLoggers.info(logger, "file:#{slot}")
              assert {:ok, request_record} = request_logger.warn.("request:#{slot}")
              {file_record, request_record}
            end)

          assert Context.current() == nil
          assert NextLoggers.current_context() == %{}
          {slot, result}
        end,
        max_concurrency: 64,
        ordered: false,
        timeout: 5_000
      )
      |> Enum.map(fn {:ok, value} -> value end)

    assert length(results) == 64

    Enum.each(results, fn {slot, {file_record, request_record}} ->
      verify_record(file_record, slot, "file:#{slot}")
      verify_record(request_record, slot, "request:#{slot}")
    end)

    assert Context.current() == nil
    assert NextLoggers.current_context() == %{}
  end

  test "nested request scopes restore exact portable, Logger, and ores-otel state after an exception" do
    outer_context = context("outer")
    outer_metadata = [outer_marker: true, request_id: "request-outer"]
    outer_log_context = %{trace_id: "outer-trace", fields: %{"scope" => "outer"}}

    Context.put(outer_context)
    Logger.reset_metadata(outer_metadata)
    restore_token = NextLoggers.put_context(outer_log_context)

    try do
      assert_raise RuntimeError, "handler exploded", fn ->
        Context.run(context("inner"), fn ->
          assert Context.current().request_id == "request-inner"
          assert NextLoggers.current_context().fields["request.id"] == "request-inner"
          raise "handler exploded"
        end)
      end

      assert Context.current() == outer_context
      assert Map.new(Logger.metadata()) == Map.new(outer_metadata)
      assert NextLoggers.current_context() == outer_log_context
    after
      NextLoggers.restore_context(restore_token)
    end
  end

  test "OTelPlug lifecycle transport failure is fail-open and restores the exact prior log context" do
    failing_logger =
      OTel.new_logger("middleware-adversarial-test",
        name: "failing-sink",
        transports: [fn _record -> raise "sink unavailable" end]
      )

    outer = %{trace_id: "outer", fields: %{"scope" => "outer"}}
    outer_restore = NextLoggers.put_context(outer)

    try do
      options = OTelPlug.init(logger: failing_logger, lifecycle: true)

      response =
        conn(:get, "/failing-sink")
        |> assign(:ores_middleware_context, context("transport"))
        |> OTelPlug.call(options)
        |> send_resp(204, "")

      assert response.status == 204
      assert response.resp_body == ""
      assert is_map(response.assigns.log)
      assert response.assigns.log == response.assigns.ores_log
      assert NextLoggers.current_context() == outer
    after
      NextLoggers.restore_context(outer_restore)
    end
  end

  test "strict OTelPlug ordering fails explicitly instead of silently creating an uncorrelated logger" do
    logger = OTel.new_logger("middleware-adversarial-test")
    options = OTelPlug.init(logger: logger, strict: true)

    assert_raise ArgumentError, ~r/must run after OresMiddleware.Plug/, fn ->
      conn(:get, "/wrong-order") |> OTelPlug.call(options)
    end

    assert NextLoggers.current_context() == %{}
  end

  test "request logger remains pinned after ambient context changes" do
    logger = OTel.new_logger("middleware-adversarial-test", name: "orders")
    pinned = OTel.request_logger(logger, context("pinned"))

    outer_restore =
      NextLoggers.put_context(%{
        trace_id: "bbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbb",
        fields: %{
          "request.id" => "request-other",
          "user.id" => "user-other",
          "tenant.id" => "tenant-other"
        }
      })

    try do
      assert {:ok, record} = pinned.info.("pinned logger")
      verify_record(record, "pinned", "pinned logger")
    after
      NextLoggers.restore_context(outer_restore)
    end
  end

  defp context(slot) do
    %{
      request_id: "request-#{slot}",
      trace_id: String.pad_leading(slot, 32, "a") |> String.slice(-32, 32),
      span_id: "0123456789abcdef",
      tenant_id: "tenant-#{slot}",
      user_id: "user-#{slot}",
      locale: "en-US",
      started_at_unix_ms: 1,
      deadline_unix_ms: 2,
      baggage: %{
        "otel.slot" => slot,
        "authorization" => "must-not-propagate",
        "cookie" => "must-not-propagate"
      }
    }
  end

  defp verify_record(record, slot, message) do
    assert record["message"] == message
    assert record["fields"]["request.id"] == "request-#{slot}"
    assert record["fields"]["user.id"] == "user-#{slot}"
    assert record["fields"]["tenant.id"] == "tenant-#{slot}"
    assert record["fields"]["otel.baggage"] == %{"otel.slot" => slot}
    assert record["traceId"] == context(slot).trace_id
    refute inspect(record) =~ "must-not-propagate"
    refute inspect(record) =~ "authorization"
    refute inspect(record) =~ "cookie"
  end
end
